cmd/soroban-cli/src/commands/contract/build.rs

```diff
@@ -229,3 +232,81 @@ impl Cmd {
         cmd.exec()
     }
 }
+
+/// Configure cargo/rustc to replace absolute paths in panic messages / debuginfo
+/// with relative paths.
+///
+/// This is required for reproducible builds.
+///
+/// This works for paths to crates in the registry. The compiler already does
+/// something similar for standard library paths and local paths. It may not
+/// work for crates that come from other sources, including the standard library
+/// compiled from source, though it may be possible to accomodate such cases in
+/// the future.
+///
+/// This in theory breaks the ability of debuggers to find source code, but
+/// since we are only targetting wasm, which is not typically run in a debugger,
+/// and stellar-cli only compiles contracts in release mode, the impact is on
+/// debugging is expected to be minimal.
+///
+/// This works by setting the `CARGO_ENCODED_RUSTFLAGS` environment variable,
+/// with appropriate `--remap-path-prefix` option. It preserves the values of an
+/// existing `CARGO_ENCODED_RUSTFLAGS` (or `RUSTFLAGS`).
+///
+/// This must be done via `RUSTFLAGS` and not as arguments to `cargo rustc`
+/// because the latter only applies to the crate directly being compiled, while
+/// `RUSTFLAGS` applies to all crates, including dependencies.
+///
+/// Note that a consequence of setting `RUSTFLAGS` via the environment is that
+/// cargo will ignore any `build.rustflags`, etc. configuration values on
+/// the local system. It may be possible to accomodate such configurations
+/// by using the cargo API directly, but that's a heavy solution.
+///
+/// This assumes that paths are Unicode and that any existing `RUSTFLAGS`
+/// variables are Unicode. Non-Unicode paths will fail to correctly perform the
+/// the absolute path replacement. Non-Unicode `RUSTFLAGS` will result in the
+/// existing `RUSTFLAGS` being ignored, though note this is also the behavior of
+/// cargo itself.
+fn set_env_to_remap_absolute_paths(cmd: &mut Command) -> Result<(), Error> {
```

Comment:
Given that this prevents use of the config `build.rustflags` and config target rustflags, maybe we should display a warning, or error, that those config settings are set. It means we wouldn't silently turn off config that someone is potentially using today. 🤔

Reply:
It would be good. If we could discover that information we could even fold it into our CARGO_ENCODED_RUSTFLAGS so we don't throw away the user's settings.

At the moment this code is run (we haven't run cargo yet) it is difficult to tell whether these are set in a file that is going to matter when we run cargo, and as far as I can tell there is no way to query cargo about what config items are going to be set when we do run a cargo command (.e.g. `cargo metadata` emits a bunch of stuff but doesn't seem to include rustflags).

We could embed the cargo crate and attempt to use it to discover this information, though the cargo API is unwieldy and not designed for external users. We could also develop a custom heuristic to search the filesystem for cargo config files approximately how cargo would do it, load the json for these keys, etc. Doable - and perhaps useful enough to publish as its own crate since it would be a slog and surely others want this type of functionality.

Even if it's not possible to query cargo about this, we could look at `CARGO_BUILD_RUSTFLAGS` and `CARGO_TARGET_wasm32-unknown-unknown_RUSTFLAGS` and slurp those in in a way that is analogous to what cargo itself would do with them if they were set. (Re-reading the cargo code we would have more compatibility with cargo if we did this, so we probably should.